kernel: I2C better driver for MT7621/MT7628

j123b567 opened the pull request:
New supported features
 - unlimited message length
 - clock stretching
 - ACK handling - i2c-detect now works - it probably fixes bug #845
 - repeated start sequence

Signed-off-by: Jan Breuer <jan.breuer@jaybee.cz>

j123b567 committed:
ramips: remove i2c driver for MT7621/MT7628/MT7688 using auto mode

This driver is now obsolate.

Signed-off-by: Jan Breuer <jan.breuer@jaybee.cz>

hitech95 commented:
Did you used the manual i2c mode?
I have checked back in time but I didn't found any documents for that!

j123b567 commented:
Yes, I'm using I2C manual mode. Just registers are documented. I have written python script to directly access I2C registers from userspace to try how it works and later write kernel driver.

hitech95 commented:
Good work. I'm not an expert so I stopped after seeing that the "auto" mode had a lot of missing function like the message length.

blogic commented:
@j123b567 so this essentially adds a new driver ? please create 2 patches,
1) remove old driver
2) add new driver
as this is quite a huge change please wrte an appropriate patch description explaining the auto vs manual mode thing etc etc

blogic commented:
and the prefix should be "ramips:"

j123b567 committed:
ramips: add i2c MT7621/MT7628/MT7688 driver using manual mode

I2C on MT7621/MT7628/MT7688 has two modes: "auto" and "manual". Old driver
uses "auto" mode which does not support propper ACK/NAK handling, clock
stratching, repeated start and was limited to 32 bytes per message.
This driver is using "manual" mode so it can support all these features.
Main advantages over old driver are:
 - unlimited message length
 - clock stretching
 - ACK/NAK handling (i2c-detect now works)
 - repeated start sequence

Signed-off-by: Jan Breuer <jan.breuer@jaybee.cz>

j123b567 commented:
I have tested this driver just on MT7688, which is the same SoC as MT7628 (except WiFi).

According to the datasheet, it should work on MT7621, because it has completely the same I2C registers, but I don't have such hardware.

yatinvasava commented:
hey it can work with any devices means it can be used for other i2c devices too ???

j123b567 commented:
@yatinvasava it is usable for example for VoCore2, Linklt Smart 7688, Onion Omega2 and Omega2+

yatinvasava commented:
Ohhh thanks... 😊😊😊😊

On Thu 5 Apr, 2018, 1:01 PM Jan Breuer, <notifications@github.com> wrote:

> @yatinvasava <https://github.com/yatinvasava> it is usable for example
> for VoCore2, Linklt Smart 7688, Onion Omega2 and Omega2+
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/831#issuecomment-378845666>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AZOwT0aQPkxJe-knlEnVYldZtqFrvyLOks5tlchggaJpZM4TA0Kv>
> .
>

neheb commented:
The commits are missing Signed-off-by. Also a comment could be useful for the driver removal commit. Something like this is now obsolete.

j123b567 commented:
@neheb It should be fixed now

neheb commented:
Kernel version 4.9 got axed from mainline and so did the patch directory.

This is a bit confusing. So the driver is the same, just modified with extra features. This is the diff I get with some tweaking.

I think it would be better to make the driver a part of the files-4.14 directory instead of a patch file. Look at the recent changes in the MMC driver for ramips for hints.

j123b567 commented:
@neheb I can change any formal representation of the driver/patches (move to dir,  etc.) but please in th meantime, can you comment the driver code itself? Test it on some other platform then me, etc? 

This change is very radical. Only the skeleton of the original driver is reused. The core part of the communication is completely different. So no, it does not just add extra features, it uses completely different approach to work with the SoC registers and to communicate with outer world.

neheb commented:
I don't have any ramips devices with anything connected to i2c, so I can't really test.

After looking into this a bit more, there is a mainline kernel driver for Mediatek i2c that can probably be adapted to replace the OpenWrt specific one. Although that's probably a longer term solution.

j123b567 commented:
@neheb I can find only i2c-mt65xx driver in the mainline which is for ARM based SoC (MT7622, MT65xx,  etc.) . It has completely different registers and for example it has DMA available for I2C.

It is not compatible in any way with MIPS based MT7621 and friends.

j123b567 commented:
I have feedback that it does not work on all similar SoCs as expected (namely mt7628), I must investigate the problem first.

mkresin commented:
@j123b567 do you have an update for us? What doesn't actually work with mt7628?

j123b567 commented:
@mkresin No, I don't have any update.

We are using this driver in production on mt7688. Maybe, i can limit this driver just for this cpu and it can replace old driver on other cpus later.

blogic commented:
@j123b567 could you try to send this upstream to linux-i2c ?

j123b567 commented:
@blogic ok, I will try to send this upstream.

blogic commented:
@j123b567 please CC me. in general the code looks good. i am reluctant to merge it as its a huge change and we dont know if it will work on all old SoC versions. I dont have the HW to test thisyet we have units with rt305x that use i2c which will likely break. carrying 2 out of tree drivers is also very cumbersome and eats up a chunk of my time. If we manage to get this upstream, it'll be backportable easily

hitech95 commented:
@j123b567 Have you been able to push this into mainline?

j123b567 commented:
@hitech95 no, but there is currently effort to do so https://marc.info/?l=linux-i2c&m=155688877527779&w=2

hitech95 commented:
@j123b567 Looks like that the patch has been merged into mainline.
@blogic Any idea when we can switch to the better driver? On mt7628 is working fine.
@mkresin I'm using it on the WRTNode2P, all fine so far

adrianschmutzler commented:
@hitech95 Maybe it's best to continue this discussion in https://github.com/openwrt/openwrt/pull/2385 ?
